### PR TITLE
C++: Recognize any non-overflowing arithmetic expression as a barrier for `cpp/uncontrolled-arithmetic`

### PR DESCRIPTION
**Pull request by MathiasVP**
This should be one of the final low-hanging fruits for this query that's motivated by real-world FPs.

(Part of https://github.com/github/codeql-c-team/issues/272.)

### CODE REVIEW

**Comment by geoffw0** [comment]
This is similar to something we did on another query recently, right?

**Comment by MathiasVP** [comment]
> This is similar to something we did on another query recently, right?

Yes. We did this in https://github.com/github/codeql/pull/5903. Once this PR is merged I'll pull the `bounded` predicate into a shared file and use it in both `cpp/uncontrolled-allocation-size`, `cpp/uncontrolled-arithmetic` and `cpp/tainted-arithmetic`.

**Comment by geoffw0** [comment]
> Once this PR is merged I'll pull the bounded predicate into a shared file and use it in both cpp/uncontrolled-allocation-size, cpp/uncontrolled-arithmetic and cpp/tainted-arithmetic.

Sounds good.

Can you point me to some LGTM projects for which this query has results?

**Comment by MathiasVP** [comment]
> Can you point me to some LGTM projects for which this query has results?

Our list of usual suspect has some results: https://lgtm.com/query/6937375284332555889/.

In terms of where this PR makes a difference, you can see an instance of this on the `python-pillow/Pillow` project if you compare the query [without this change](https://lgtm.com/query/1983045646580655394/) to the query [after this change](https://lgtm.com/query/1983045646580655394/).

The other projects in that run are similar cases which sadly need a more syntactic barrier which I haven't added yet.

File: cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql:69]: If I'm understanding correctly, this can in certain cases fall short of the specification "greatly reduces the range of possible values" - to give an extreme example the following is now a barrier:
```
x = x + 0; // this can't overflow
```
However this is pretty silly, in practice we expect the overwhelming majority of arithmetic either can cause an overflow (`x = x + 1`), or greatly narrows the range of values (`x = x / RAND_MAX`) ... or is working on an already constrained value (hopefully safely).

If my above understanding is accurate, this is a somewhat heuristic approach, but I think I'm happy with it.

File: cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql:67]: What about `AssignArithmeticOperation`?

File: cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql:62]: In your new case `e` is the operation itself, not an operand (I don't think this is bad aside from not matching the QLDoc).

**Review by geoffw0 — COMMENTED** [review]
A few comments.

Can we get rid of any of the `boundedDiv` cases in `bounded` now?

File: cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql:69]: Indeed. This whole predicate uses a bunch of heuristics.

File: cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql]: Good catch! Fixed in bb6a2d3845a43b95c8dc23195857977e06087448.

File: cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql]: Good point. I've added this one in bb6a2d3845a43b95c8dc23195857977e06087448.

**Comment by MathiasVP** [comment]
> Can we get rid of any of the `boundedDiv` cases in `bounded` now?

Unfortunately not. `SimpleRangeAnalysis` only handles division and right shifts when one of the operations is a constant. Whereas `boundedDiv` removes them regardless of whether or not they are constants.

**Comment by MathiasVP** [comment]
@geoffw0 I believe all your concerns have been addressed.

**Review by geoffw0 — APPROVED** [review]
Yep, thanks for the fixes, merging...

**Pull request merged**
